src/analysis/processing/qgsalgorithmdownloadvectortiles.cpp

```diff
@@ -0,0 +1,225 @@
+  }
+  writer->setMetadataValue( "format", "pbf" );
+  writer->setMetadataValue( "name",  vtLayer->name() );
+  writer->setMetadataValue( "minzoom", QString::number( vtLayer->sourceMinZoom() ) );
+  writer->setMetadataValue( "maxzoom", QString::number( maxZoomLevel ) );
+  writer->setMetadataValue( "crs",  tileMatrixSet.rootMatrix().crs().authid() );
+  try
+  {
+    QgsCoordinateTransform ct( tileMatrixSet.rootMatrix().crs(), QgsCoordinateReferenceSystem( "EPSG:4326" ), context.transformContext() );
+    ct.setBallparkTransformsAreAppropriate( true );
+    QgsRectangle wgsExtent = ct.transform( extent );
+    QString boundsStr = QString( "%1,%2,%3,%4" )
+                        .arg( wgsExtent.xMinimum() ).arg( wgsExtent.yMinimum() )
+                        .arg( wgsExtent.xMaximum() ).arg( wgsExtent.yMaximum() );
+    writer->setMetadataValue( "bounds", boundsStr );
+  }
+  catch ( const QgsCsException & )
+  {
+    // bounds won't be written (not a problem - it is an optional value)
+  }
+
+  QgsProcessingMultiStepFeedback multiStepFeedback( maxZoomLevel + 1, feedback );
+
+  std::unique_ptr<QgsVectorTileLoader> loader;
+  QList<QgsVectorTileRawData> rawTiles;
+
+  QMap<int, QgsTileRange>::const_iterator it = tileRanges.constBegin();
+  while ( it != tileRanges.constEnd() )
+  {
+    if ( feedback->isCanceled() )
+      break;
+
+    multiStepFeedback.setCurrentStep( it.key() );
+
+    QgsTileMatrix tileMatrix = tileMatrixSet.tileMatrix( it.key() );
+    tileCount = ( it.value().endColumn() - it.value().startColumn() + 1 ) * ( it.value().endRow() - it.value().startRow() + 1 );
+
+    const QPointF viewCenter = tileMatrix.mapToTileCoordinates( extent.center() );
+
+    long long tileNumber = 0;
+    if ( !provider->supportsAsync() )
+    {
+      rawTiles = QgsVectorTileLoader::blockingFetchTileRawData( provider, tileMatrixSet, viewCenter, it.value(), it.key(), &multiStepFeedback );
+      for ( const QgsVectorTileRawData &rawTile : std::as_const( rawTiles ) )
+      {
+        if ( feedback->isCanceled() )
+          break;
+
+        writeTile( writer.get(), rawTile );
+        multiStepFeedback.setProgress( 100.0 * ( tileNumber++ ) / tileCount );
+      }
+    }
+    else
+    {
+      loader.reset( new QgsVectorTileLoader( provider, tileMatrixSet, it.value(), it.key(), viewCenter, &multiStepFeedback, Qgis::RendererUsage::Export ) );
+      QObject::connect( loader.get(), &QgsVectorTileLoader::tileRequestFinished, loader.get(), [ this, &writer, &multiStepFeedback, &tileNumber, &tileCount ]( const QgsVectorTileRawData & rawTile )
+      {
+        writeTile( writer.get(), rawTile );
+        multiStepFeedback.setProgress( 100.0 * ( tileNumber++ ) / tileCount );
+      } );
+      loader->downloadBlocking();
+    }
+
+    ++it;
+  }
+
+  if ( context.willLoadLayerOnCompletion( outputFile ) )
+  {
+    context.layerToLoadOnCompletionDetails( outputFile ).setPostProcessor( new SetStylePostProcessor( vtLayer ) );
```

Comment:
There's potentially a threading issue here if vtLayer is deleted before the post processor is run. Could you move the exportNamedStyle call from the post processor to the prepareAlgorithm step here, and then case the exported style document to the post processor instead of the input layer?

Ideally we'd get rid of all use of vtLayer from processAlgorithm

Reply:
Should be fixed, thanks for the suggestion.